File: binding/java/src/main/java/org/lionsoul/ip2region/IndexBlock.java
package org.lionsoul.ip2region;

/**
 * item index class
 * 
 * @author chenxin<chenxin619315@gmail.com>
*/
public class IndexBlock 
{
    private static int LENGTH = 12;
    
    /**
     * start ip address 
    */
    private long startIp;
    
    /**
     * end ip address 
    */
    private long endIp;
    
    /**
     * data ptr and data length 
    */
    private int dataPtr;
    
    /**
     * data length 
    */
    private int dataLen;
    
    public IndexBlock(long startIp, long endIp, int dataPtr, int dataLen)
    {
        this.startIp = startIp;
        this.endIp = endIp;
        this.dataPtr = dataPtr;
        this.dataLen = dataLen;
    }

    public long getStartIp()
    {
        return startIp;
    }

    public IndexBlock setStartIp(long startIp)
    {
        this.startIp = startIp;
        return this;
    }

    public long getEndIp()
    {
        return endIp;
    }

    public IndexBlock setEndIp(long endIp)
    {
        this.endIp = endIp;
        return this;
    }

    public int getDataPtr()
    {
        return dataPtr;
    }

    public IndexBlock setDataPtr(int dataPtr)
    {
        this.dataPtr = dataPtr;
        return this;
    }

    public int getDataLen()
    {
        return dataLen;
    }

    public IndexBlock setDataLen(int dataLen)
    {
        this.dataLen = dataLen;
        return this;
    }
    
    public static int getIndexBlockLength()
    {
        return LENGTH;
    }
    
    /**
     * get the bytes for storage
     * 
     * @return    byte[]
    */
    public byte[] getBytes()
    {
        /*
         * +------------+-----------+-----------+
         * | 4bytes        | 4bytes    | 4bytes    |
         * +------------+-----------+-----------+
         *  start ip      end ip      data ptr + len 
        */
        byte[] b = new byte[12];
        
        Util.writeIntLong(b, 0, startIp);    //start ip
        Util.writeIntLong(b, 4, endIp);        //end ip
        
        //write the data ptr and the length
        long mix = dataPtr | ((dataLen << 24) & 0xFF000000L);
        Util.writeIntLong(b, 8, mix);
        
        return b;
    }
}

File: maker/csharp/DbMaker/IndexBlock.cs
namespace DbMaker
{
    /// <summary>
    ///     item index class
    /// </summary>
    public class IndexBlock
    {
        private static int LENGTH = 12;

        public IndexBlock(long startIp, long endIp, int dataPtr, int dataLen)
        {
            StartIp = startIp;
            EndIp = endIp;
            DataPtr = dataPtr;
            DataLen = dataLen;
        }

        /**
         * start ip address
         */
        public long StartIp { get; set; }

        /**
         * end ip address
         */
        public long EndIp { get; set; }

        /**
         * data ptr and data length
         */
        public int DataPtr { get; set; }

        /**
         * data length
         */
        public int DataLen { get; set; }

        public IndexBlock SetStartIp(long startIp)
        {
            StartIp = startIp;
            return this;
        }

        public IndexBlock SetEndIp(long endIp)
        {
            EndIp = endIp;
            return this;
        }

        public IndexBlock SetDataPtr(int dataPtr)
        {
            DataPtr = dataPtr;
            return this;
        }

        public IndexBlock SetDataLen(int dataLen)
        {
            DataLen = dataLen;
            return this;
        }

        public static int GetIndexBlockLength()
        {
            return LENGTH;
        }

        /**
     * get the bytes for storage
     * 
     * @return    byte[]
    */
        public byte[] GetBytes()
        {
            /*
             * +------------+-----------+-----------+
             * | 4bytes        | 4bytes    | 4bytes    |
             * +------------+-----------+-----------+
             *  start ip      end ip      data ptr + len 
            */
            var b = new byte[12];

            Util.writeIntLong(b, 0, StartIp); //start ip
            Util.writeIntLong(b, 4, EndIp); //end ip

            //write the data ptr and the length
            var mix = DataPtr | DataLen << 24 & 0xFF000000L;
            Util.writeIntLong(b, 8, mix);

            return b;
        }
    }
}